### PR TITLE
fix(migrations):  make transactions match SaaS SNS-1991

### PR DESCRIPTION
**Pull request by dbanda**
This PR attempts to align the transactions table schemas with the schemas in SaaS. It does the following:

1. Touches up some of the existing migrations to display columns in the same order at SaaS. This makes no functional difference to the schemas but by displaying columns in a consistent order it allows our diff tool to more easily catch when schemas are different.
2. Adds a migration step to add the indexes to the transactions table
3. Adds a step to add the compression codecs to columns to match SaaS. Notably, this skips the event_id and span_id columns  because Clickhouse 20 does not allow modifying key columns
4. Adds context_hash_map to the transactions dist tables and updates the other hash columns to remove the materialized modifier to match SaaS

### CODE REVIEW

**Review by evanh — COMMENTED** [review]
Are any of the `ModifyColumn` ops blocking in some way? When we tried to modify the TTL of a transactions column it brought the entire transactions cluster to a stop while it applied the change.

I ask because I don't want one of the self hosted clients to run these and have their app stop working.

File: snuba/snuba_migrations/transactions/0018_transactions_add_indexes.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                ),
+                after="contexts.value",
+                target=operations.OperationTarget.LOCAL,
+            ),
+        ]
+
+        # apply the indexes after the minmax_timestamp index and have them follow each other
+        # one after the other
+        index_names = [index[0] for index in indexes]
+        for index, after_col in zip(indexes, ["minmax_timestamp"] + index_names[:-1]):
```
[the remark this answers is not in the document]
Comment (dbanda) [reply snuba/snuba_migrations/transactions/0018_transactions_add_indexes.py:58]: The `minimax_timestamp` index was already added by previous migrations.

**Comment by dbanda** [comment]
> Are any of the `ModifyColumn` ops blocking in some way? When we tried to modify the TTL of a transactions column it brought the entire transactions cluster to a stop while it applied the change.
> 
> I ask because I don't want one of the self hosted clients to run these and have their app stop working.

According to the Clickhouse docs ADD/DROP INDEX are "lightweight in a sense that they only change metadata or remove files"

File: snuba/snuba_migrations/transactions/0018_transactions_add_indexes.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            operations.AddColumn(
+                storage_set=StorageSetKey.TRANSACTIONS,
+                table_name="transactions_local",
+                column=Column(
+                    "_contexts_hash_map",
+                    Array(UInt(64), Modifiers(materialized=CONTEXTS_HASH_MAP_COLUMN)),
+                ),
+                after="contexts.value",
+                target=operations.OperationTarget.LOCAL,
+            ),
```
Comment (nikhars) [inline snuba/snuba_migrations/transactions/0018_transactions_add_indexes.py:52]: Please add the contexts hash map for the local and distributed tables in the same migration. Currently for local tables you are defining it in 0018 and for dist tables in 0020. A failure in migration 0019 would leave the system in a non-consistent state where local tables have a column while the dist tables don't.

**Comment by nikhars** [comment]
> 



> > Are any of the `ModifyColumn` ops blocking in some way? When we tried to modify the TTL of a transactions column it brought the entire transactions cluster to a stop while it applied the change.
> > I ask because I don't want one of the self hosted clients to run these and have their app stop working.
> 
> According to the Clickhouse docs ADD/DROP INDEX are "lightweight in a sense that they only change metadata or remove files"

That's correct. We should set blocking to True for migrations which apply TTL. Not necessary for any migrations being applied in this PR

File: snuba/snuba_migrations/transactions/0018_transactions_add_indexes.py (hunk 2 above)
Comment (dbanda) [reply snuba/snuba_migrations/transactions/0018_transactions_add_indexes.py]: sounds good. moved both to 0019_transactions_add_indexes_and_context_hash

File: snuba/snuba_migrations/transactions/0018_transactions_add_indexes.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (evanh) [reply snuba/snuba_migrations/transactions/0018_transactions_add_indexes.py]: OK, then why is `minmax_timestamp` in the list of indexes?

**Comment by evanh** [comment]
> According to the Clickhouse docs ADD/DROP INDEX are "lightweight in a sense that they only change metadata or remove files"

OK perfect, thanks for checking that.

File: snuba/snuba_migrations/transactions/0018_transactions_add_indexes.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nikhars) [reply snuba/snuba_migrations/transactions/0018_transactions_add_indexes.py]: I have the same question as Evan

File: snuba/snuba_migrations/transactions/0020_transactions_add_codecs.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+class Migration(migration.ClickhouseNodeMigration):
+    """
+    This migrations adds the codecs to columns to match what is in SaaS.
+    """
+
+    blocking = False
+
+    def forwards_ops(self) -> Sequence[operations.SqlOperation]:
+        return [
+            # Modifying the codec of a key column is not supported in Clickhouse 20.
```
Comment (MeredithAnya) [inline snuba/snuba_migrations/transactions/0020_transactions_add_codecs.py:18]: nit: I think comment and the comment below could just go in the doc string.

File: snuba/snuba_migrations/transactions/0018_transactions_add_indexes.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (MeredithAnya) [reply snuba/snuba_migrations/transactions/0018_transactions_add_indexes.py]: my understanding is that `minmax_timestamp` is just used for the `after_col` right?

[the remark this answers is not in the document]
Comment (dbanda) [reply snuba/snuba_migrations/transactions/0018_transactions_add_indexes.py]: that's correct. remove the index from the list of indexes to make that clear.

File: snuba/datasets/storages/tags_hash_map.py [hunk 4]
```diff
@@ -10,6 +10,12 @@
     "tags.key, tags.value)"
 )
 
+CONTEXTS_HASH_MAP_COLUMN = (
+    "arrayMap((k, v) -> cityHash64(concat("
+    "replaceRegexpAll(k, '(\\\\=|\\\\\\\\)', '\\\\\\\\\\\\1'), '=', v)), "
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/datasets/storages/tags_hash_map.py:15]: there are multiple encodings there. Each one doubles the backslashes as it adds escaping:
- python string  (double quotes)
- clickhouse string (single quotes)
- regexp encoding.

You are replacing this string `(=|\)`. 
In order to make it a regexp: `(\=|\\)`
Now it is encoded in the clickhosue statement as a string: `(\\=|\\\\)`
Now it has to be encoded as a python string: `(\\\\=|\\\\\\\\)`

I don't know what humanity did wrong to deserve escaping, but that's where we are.

**Review by MeredithAnya — APPROVED** [review]

**Pull request merged**